internal/dataplane/kongstate/kongstate.go

```diff
@@ -342,21 +344,40 @@ func buildPlugins(
 			}
 		}
 
+		usedInstanceNames := sets.New[string]()
 		for _, rel := range relations.GetCombinations() {
 			plugin := plugin.DeepCopy()
+			var sha [32]byte
 			// ID is populated because that is read by decK and in_memory
 			// translator too
 			if rel.Service != "" {
 				plugin.Service = &kong.Service{ID: kong.String(rel.Service)}
+				sha = sha256.Sum256([]byte(rel.Service))
 			}
 			if rel.Route != "" {
 				plugin.Route = &kong.Route{ID: kong.String(rel.Route)}
+				sha = sha256.Sum256([]byte(rel.Route))
 			}
 			if rel.Consumer != "" {
 				plugin.Consumer = &kong.Consumer{ID: kong.String(rel.Consumer)}
+				sha = sha256.Sum256([]byte(rel.Consumer))
 			}
 			if rel.ConsumerGroup != "" {
 				plugin.ConsumerGroup = &kong.ConsumerGroup{ID: kong.String(rel.ConsumerGroup)}
+				sha = sha256.Sum256([]byte(rel.ConsumerGroup))
```

Comment:
Although it is very unlikely to happen, could `Consumer` and `ConsumerGroup` have the same name? I suggest to add some bits annotating the type of attached entity, like `service-***`, `route-***`, `consumer-***` or `consumer-group-***`.